In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("Titanic-Dataset.csv")

In [3]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


In [4]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


## Data Quality Report

### Dataset Information

The dataset information was examined using `df.info()` to understand the number of rows, columns, data types, and identify missing values before performing data cleaning.

In [5]:
df.isnull().sum()


PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## Missing Values

The dataset contains missing values in three columns:

- **Age:** 177 missing values
- **Cabin:** 687 missing values
- **Embarked:** 2 missing values

These missing values will be handled in the data cleaning process using suitable techniques.

In [6]:
df.duplicated().sum()

np.int64(0)

## Duplicate Rows

The dataset was checked for duplicate records using `df.duplicated().sum()`.

Result:
- Duplicate Rows: 0

No duplicate records were found in the dataset, so no duplicate removal is required.

In [7]:
df.dtypes


PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Cabin              str
Embarked           str
dtype: object

## Data Types

The dataset contains integer, float, and string (object) data types.

- Integer columns: PassengerId, Survived, Pclass, SibSp, Parch
- Float columns: Age, Fare
- String (Object) columns: Name, Sex, Ticket, Cabin, Embarked

The data types appear appropriate for analysis. No immediate data type corrections are required.

In [8]:
(df.isnull().sum() / len(df)) * 100

PassengerId     0.000000
Survived        0.000000
Pclass          0.000000
Name            0.000000
Sex             0.000000
Age            19.865320
SibSp           0.000000
Parch           0.000000
Ticket          0.000000
Fare            0.000000
Cabin          77.104377
Embarked        0.224467
dtype: float64

# Missing Data Handling

I checked the percentage of missing values in each column before cleaning the data. This helped me decide whether to fill the missing values or remove the column.

In [9]:

df["Age"] = df["Age"].fillna(df["Age"].median())

df["Embarked"] = df["Embarked"].fillna(df["Embarked"].mode()[0])


df.drop("Cabin", axis=1, inplace=True)


In [10]:
df.isnull().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Embarked       0
dtype: int64

## Handling Missing Values

I filled the missing values in the Age column using the median and the Embarked column using the most common value (mode). Since the Cabin column had too many missing values, I removed it from the dataset.

In [11]:
df.to_csv("Titanic_Cleaned.csv", index=False)

In [12]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,S


In [13]:
df.shape

(891, 11)

In [14]:
df.to_csv("Titanic_Cleaned.csv", index=False)


## Final Dataset

After cleaning the dataset, all missing values were handled successfully. The Cabin column was removed because it contained a large number of missing values. The cleaned dataset was saved as **Titanic_Cleaned.csv** and is now ready for analysis.

In [15]:
df.duplicated().sum()

np.int64(0)

## Duplicate Rows

I checked the dataset for duplicate rows using `df.duplicated().sum()`. No duplicate rows were found, so no records were removed.

In [16]:
df["Sex"] = df["Sex"].str.lower()
df["Embarked"] = df["Embarked"].str.upper()

## Standardizing Text

I standardized the text values to keep the data consistent. The values in the **Sex** column were converted to lowercase, and the **Embarked** column was converted to uppercase.

In [17]:
Q1 = df["Fare"].quantile(0.25)
Q3 = df["Fare"].quantile(0.75)

IQR = Q3 - Q1

lower_limit = Q1 - 1.5 * IQR
upper_limit = Q3 + 1.5 * IQR

outliers = df[(df["Fare"] < lower_limit) | (df["Fare"] > upper_limit)]

outliers


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C
27,28,0,1,"Fortune, Mr. Charles Alexander",male,19.0,3,2,19950,263.0000,S
31,32,1,1,"Spencer, Mrs. William Augustus (Marie Eugenie)",female,28.0,1,0,PC 17569,146.5208,C
34,35,0,1,"Meyer, Mr. Edgar Joseph",male,28.0,1,0,PC 17604,82.1708,C
52,53,1,1,"Harper, Mrs. Henry Sleeper (Myna Haxtun)",female,49.0,1,0,PC 17572,76.7292,C
...,...,...,...,...,...,...,...,...,...,...,...
846,847,0,3,"Sage, Mr. Douglas Bullen",male,28.0,8,2,CA. 2343,69.5500,S
849,850,1,1,"Goldenberg, Mrs. Samuel L (Edwiga Grabowska)",female,28.0,1,0,17453,89.1042,C
856,857,1,1,"Wick, Mrs. George Dennick (Mary Hitchcock)",female,45.0,1,1,36928,164.8667,S
863,864,0,3,"Sage, Miss. Dorothy Edith ""Dolly""",female,28.0,8,2,CA. 2343,69.5500,S


## Outlier Detection

I used the IQR (Interquartile Range) method to detect outliers in the **Fare** column. The outliers were identified for analysis but were not removed because they may represent valid passenger fares.

In [18]:
df.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name               str
Sex                str
Age            float64
SibSp            int64
Parch            int64
Ticket             str
Fare           float64
Embarked           str
dtype: object

## Data Types

I checked the data types of all columns. The existing data types were appropriate for analysis, so no changes were required.

In [19]:
summary = {
    "Rows": [891, df.shape[0]],
    "Columns": [12, df.shape[1]],
    "Missing Values": [866, df.isnull().sum().sum()],
    "Duplicate Rows": [0, df.duplicated().sum()]
}

import pandas as pd
comparison = pd.DataFrame(summary, index=["Before Cleaning", "After Cleaning"])
comparison


,Rows,Columns,Missing Values,Duplicate Rows
Before Cleaning,891,12,866,0
After Cleaning,891,11,0,0


In [20]:
df.to_csv("Titanic_Cleaned.csv", index=False)

